ManPages/tsManPagesLibGUI/tsWxButton.man

Help on module tsWxButton:

NAME
    tsWxButton

FILE
    ./tsLibGUI/tsWxPkg/src/tsWxButton.py

DESCRIPTION
    tsWxButton.py - Class to establish a button which is a control
    that contains a text string. It is one of the most common elements
    of a GUI. It may be placed on a dialog box or panel, or indeed
    almost any other window.

CLASSES
    tsWxControl.Control(tsWxWindow.Window)
        Button
    
    class Button(tsWxControl.Control)
     |  A button is a control that contains a text string, and is one of the
     |  most common elements of a GUI. It may be placed on a dialog box or
     |  panel, or indeed almost any other window.
     |  
     |  Method resolution order:
     |      Button
     |      tsWxControl.Control
     |      tsWxWindow.Window
     |      tsWxEvtHandler.EvtHandler
     |      tsWxObject.Object
     |      __builtin__.object
     |  
     |  Methods defined here:
     |  
     |  Create(self, parent, id=-1, label='', pos=(-1, -1), size=(-1, -1), style=0, validator=None, name='button', pixels=True)
     |      Create the GUI Button for 2-phase creation.
     |  
     |  GetDefaultSize(self)
     |      Returns the default button size for this platform.
     |  
     |  SetDefault(self)
     |      This sets the button to be the default item for the panel or
     |      dialog box.
     |  
     |  __init__(self, parent, id=-1, label='', pos=(-1, -1), size=(-1, -1), style=0, validator=None, name='button', useClientArea=True)
     |      Create and show a button. The preferred way to create standard buttons
     |      is to use a standard ID and an empty label. In this case wxWigets will
     |      automatically use a stock label that corresponds to the ID given.
     |      These labels may vary across platforms as the platform itself will
     |      provide the label if possible. In addition, the button will be
     |      decorated with stock icons under GTK+ 2.
     |  
     |  tsButtonLayout(self, parent, pos, size, style, name)
     |      Calculate position and size of button based upon arguments.
     |  
     |  tsOnLeftClick(self, evt)
     |  
     |  tsShow(self)
     |      Create and update Button specific native GUI.
     |  
     |  tsUpdate(self)
     |      Draw the actual features of the Button.
     |  
     |  ----------------------------------------------------------------------
     |  Static methods defined here:
     |  
     |  GetClassDefaultAttributes(variant=0)
     |      Get the default attributes for this class. This is useful if you want
     |      to use the same font or colour in your own control as in a standard
     |      control -- which is a much better idea than hard coding specific
     |      colours or fonts which might look completely out of place on the
     |      user system, especially if it uses themes.
     |      
     |      The variant parameter is only relevant under Mac currently and is
     |      ignored under other platforms. Under Mac, it will change the size of
     |      the returned font. See wx.Window.SetWindowVariant for more about this.
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from tsWxControl.Control:
     |  
     |  Command(self, event)
     |      Simulates the effect of the user issuing a command to the item.
     |  
     |  GetAlignment(self)
     |      Get the control alignment (left/right/centre, top/bottom/centre)
     |  
     |  GetLabelText(self)
     |      Get just the text of the label, without mnemonic characters ("&")
     |  
     |  ----------------------------------------------------------------------
     |  Data descriptors inherited from tsWxControl.Control:
     |  
     |  Alignment
     |      Get the control alignment (left/right/centre, top/bottom/centre)
     |  
     |  LabelText
     |      Get just the text of the label, without mnemonic characters ("&")
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from tsWxWindow.Window:
     |  
     |  AcceptsFocus(self)
     |      This method may be overridden in the derived classes to return false
     |      to indicate that this control does not accept input at all
     |  
     |  AcceptsFocusFromKeyboard(self)
     |      This method may be overridden in the derived classes to return false
     |      to indicate that while this control can, in principle, have focus if
     |      the user clicks it with the mouse, it should not be included in the
     |      TAB traversal chain when using the keyboard.
     |  
     |  AcceptsFocusRecursively(self)
     |      Overridden to indicate whether this window or one of its children
     |      accepts focus.
     |  
     |  AddChild(self, child)
     |      Adds a child window. This is called automatically by window creation
     |      functions so should not be required by the application programmer.
     |  
     |  AddPendingEvent(self, event)
     |      See ProcessEvent() for more info about why you should not use
     |      this function and the reason for making this function protected
     |      in wxWindow.
     |      
     |      Reimplemented from wxEvtHandler.
     |  
     |  AdjustForLayoutDirection(self, x, width, widthTotal)
     |      Mirror coordinates for RTL layout if this window uses it and
     |      if the mirroring is not done automatically like Win32.
     |  
     |  AdjustForParentClientOriginXY(self, x, y, sizeFlags)
     |  
     |  AlwaysShowScrollbars(self)
     |      Call this function to force one or both scrollbars to be always
     |      shown, even if the window is big enough to show its entire contents
     |      without scrolling.
     |  
     |  AssociateHandle(self, handle)
     |      Associate the window with a new native handle.
     |  
     |  CacheBestSize(self, size)
     |      Sets the cached best size value.
     |  
     |  CanScroll(self, orient)
     |      Returns True if this window can have a scroll bar in this
     |      orientation.
     |  
     |  CanSetTransparent(self)
     |      Returns true if the system supports transparent windows and
     |      calling SetTransparent() may succeed.
     |      
     |      If this function returns false, transparent windows are definitely
     |      not supported by the current system.
     |      
     |      Reimplemented in wxTopLevelWindow.
     |      
     |      NOTE: On X-windows systems the X server must have the composite
     |      extension loaded, and there must be a composite manager program
     |      (such as xcompmgr) running.
     |  
     |  CaptureMouse(self)
     |      Directs all mouse input to this window.
     |      
     |      Call ReleaseMouse() to release the capture.
     |      
     |      Note that wxWindows maintains the stack of windows having captured
     |      the mouse and when the mouse is released the capture returns to the
     |      window which had had captured it previously and it is only really
     |      released if there were no previous window. In particular, this means
     |      that you must release the mouse as many times as you capture it,
     |      unless the window receives the wx.MouseCaptureLostEvent event.
     |      
     |      Any application which captures the mouse in the beginning of some
     |      operation must handle wx.MouseCaptureLostEvent and cancel this
     |      operation when it receives the event. The event handler must not
     |      recapture mouse.
     |  
     |  Center(self, direction=12)
     |      Centers the window. The parameter specifies the direction for
     |      cetering, and may be wx.HORIZONTAL, wx.VERTICAL or wx.BOTH. It may
     |      also include wx.CENTER_ON_SCREEN flag if you want to center the
     |      window on the entire screen and not on its parent window. If it is
     |      a top-level window and has no parent then it will always be centered
     |      relative to the screen.
     |  
     |  CenterOnParent(self, dir=12)
     |      Center with respect to the the parent window.
     |  
     |  CenterOnScreen(self, direction=12)
     |      Center the window on screen.
     |  
     |  Centre(self, direction=12)
     |      Centers the window. The parameter specifies the direction for
     |      cetering, and may be wx.HORIZONTAL, wx.VERTICAL or wx.BOTH. It may
     |      also include wx.CENTER_ON_SCREEN flag if you want to center the
     |      window on the entire screen and not on its parent window. If it is
     |      a top-level window and has no parent then it will always be centered
     |      relative to the screen.
     |  
     |  CentreOnParent(self, dir=12)
     |      Center with respect to the the parent window.
     |  
     |  CentreOnScreen(self, direction=12)
     |      Center the window on screen.
     |  
     |  ClearBackground(self)
     |      Clears the window by filling it with the current background colour.
     |      
     |      Does not cause an erase background event to be generated.
     |      
     |      Notice that this uses wxClientDC to draw on the window and the
     |      results of doing it while also drawing on wxPaintDC for this
     |      window are undefined. Hence this method should not be used
     |      from EVT_PAINT handlers, just use wxDC::Clear() on the wxPaintDC
     |      you already use there instead.
     |  
     |  ClientToScreen(self, pt)
     |      Converts to screen coordinates from coordinates relative to this
     |      window.
     |  
     |  ClientToScreenXY(self, x, y)
     |      Converts to screen coordinates from coordinates relative to this
     |      window.
     |  
     |  ClientToWindowSize(self, size)
     |      Converts client area size size to corresponding window size.
     |  
     |  Close(self, force=True)
     |      This function simply generates a wxCloseEvent whose handler usually
     |      tries to close the window.
     |      
     |      It does not close the window itself, however.
     |      
     |      Parameters:
     |      force   false if the windows close handler should be able to veto
     |      the destruction of this window, true if it cannot.
     |      
     |      Remarks:
     |      Close calls the close handler for the window, providing an
     |      opportunity for the window to choose whether to destroy the window.
     |      Usually it is only used with the top level windows (wxFrame and
     |      wxDialog classes) as the others are not supposed to have any special
     |      OnClose() logic. The close handler should check whether the window
     |      is being deleted forcibly, using wxCloseEvent::CanVeto, in which
     |      case it should destroy the window using wxWindow::Destroy. Note
     |      that calling Close does not guarantee that the window will be
     |      destroyed; but it provides a way to simulate a manual close of
     |      a window, which may or may not be implemented by destroying the
     |      window. The default implementation of wxDialog::OnCloseWindow
     |      does not necessarily delete the dialog, since it will simply
     |      simulate an wxID_CANCEL event which is handled by the appropriate
     |      button event handler and may do anything at all. To guarantee
     |      that the window will be destroyed, call wxWindow::Destroy instead.
     |  
     |  ConvertDialogPixels(self, pt)
     |      Converts a point or size from dialog units to pixels.
     |      
     |      For the x dimension, the dialog units are multiplied by the
     |      fixed point character width. For the y dimension, the dialog
     |      units are multiplied by the fixed point character height.
     |      
     |      Remarks:
     |      Dialog units are used for maintaining a dialogs proportions

Documents/NEWS.txt
#-----------------------------------------------------------
#"Time-stamp: <07/08/2015  4:48:41 AM rsg>"
#-----------------------------------------------------------

========================= NEWS.txt =========================

   +----+----+  TeamSTARS "tsWxGTUI_PyVx" Toolkit
   | ts | Wx |      with Python 2x & Python 3x based
   +----+----+         Command Line Interface (CLI)
   | G T U I |      and "Curses"-based "wxPython"-style, 
   +---------+         Graphical-Text User Interface (GUI)
 
   Get that cross-platform, pixel-mode "wxPython" feeling
   on character-mode 8-/16-color (xterm-family) & non-color
   (vt100-family) terminals and terminal emulators.

   You can find this and other plain-text files in the
   Toolkit subdirectory named:

       "./<Toolkit Recipient's Repository>/Documents".

       <Your Working Repository>
       (e.g. "tsWxGTUI_PyVx_Repository") 
         |
         |
         +-- ["Documents"]

=========================== News ===========================

The pre-alpha release of the TeamSTARS "tsWxGTUI_PyVx-0.0.0"
Toolkit is now available. It includes the following two
versions abd their associated variants (Python Distribution
Utilities ("Distutils") installable "Site-Packages" and non-
installable "Developer-Sandboxes" for enhancement and bug
fix development and experimentation without interference
with the already installed Site-Packages):

1. The TeamSTARS "tsWxGTUI_Py2x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   second generation Python 2.x programming language which
   is widely available.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. The final 2.x
   version 2.7 release came out in mid-2010, with a state-
   ment of extended support for this end-of-life release.
   The 2.x branch will see no new major releases after that.

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 2.0.0-2.7.9.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 2.6.4-2.7.9.

2. The TeamSTARS "tsWxGTUI_Py3x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   third generation Python 3.x programming language which
   is growing in availability.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. 3.x is under
   active development and has already seen over five years
   of stable releases, including version 3.3 in 2012 and
   3.4 in 2014. This means that all recent standard library
   improvements, for example, are only available by default
   in Python 3.x.

   Guido van Rossum (the original creator of the Python
   language) decided to clean up Python 2.x properly,
   with less regard for backwards compatibility than is
   the case for new releases in the 2.x range. The most
   drastic improvement is the better Unicode support
   (with all text strings being Unicode by default) as
   well as saner bytes/Unicode separation.

   Besides, several aspects of the core language (such as
   print and exec being statements, integers using floor
   division) have been adjusted to be easier for newcomers
   to learn and to be more consistent with the rest of the
   language, and old cruft has been removed (for example,
   all classes are now new-style, "range()" returns a memory
   efficient iterable, not a list as in 2.x).

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 3.0.0-3.4.2.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 3.1.5-3.4.2.

======================= End-Of-File ========================


SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-3x/tsWxGTUI_Py3x/tsLibGUI/tsWxPkg/src/tsWxButton.py
#! /usr/bin/env python
# "Time-stamp: <10/25/2013  6:26:34 AM rsg>"
'''
tsWxButton.py - Class to establish a button which is a control
that contains a text string. It is one of the most common elements
of a GUI. It may be placed on a dialog box or panel, or indeed
almost any other window.
'''
#################################################################
#
# File: tsWxButton.py
#
# Purpose:
#
#    Class to establish a button which is a control that contains
#    a text string. It is one of the most common elements of a
#    GUI. It may be placed on a dialog box or panel, or indeed
#    almost any other window.
#
# Usage (example):
#
#    # Import
#
#    from tsWxButton import Button
#
# Requirements:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Capabilities:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Limitations:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Notes:
#
#    None
#
# Methods:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Classes:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Modifications:
#
#    2011/12/26 rsg Added logic to tsButtonLayout to inhibit
#                   operation once self.ts_Handle has been
#                   assigned a curses window identifier.
#
#    2012/02/21 rsg Modified tsOnLeftClick to invoke
#                   tsProcessEventTable.
#
#    2012/03/09 rsg Added missing objectId assignment to
#                   tsOnLeftClick.
#
#    2012/07/28 rsg Remove triggeringMouseTuple argument from
#                   self.tsProcessEventTables and from
#                   self.tsProcessSelectedEventTable since
#                   event now contains value in its EventData.
#                   Add evt argument to tsOnLeftClick.
#
#    2013/07/04 rsg Modified tsOnLeftClick to disable logic for
#                   mapping EVT_COMMAND_LEFT_CLICK into:
#
#                   EVT_CLOSE       # [X]-Button Click (Frame)
#                   EVT_HELP        # [?]-Button Click (Dialog)
#                   EVT_ICONIZE     # [_]-Button Click (Frame)
#                   EVT_MAXIMIZE    # [Z]-Button Click (Frame)
#                   EVT_RESTOREDOWN # [z]-Button Click (Frame)
#
#                   Also, disabled imports of unused events.
#
# ToDo:
#
#    None.
#
#################################################################

__title__     = 'tsWxButton'
__version__   = '1.5.0'
__date__      = '07/18/2013'
__authors__   = 'Richard S. Gordon'
__copyright__ = 'Copyright (c) 2007-2013 ' + \
                '%s.\n\t\tAll rights reserved.' % __authors__
__license__   = 'GNU General Public License, ' + \
                'Version 3, 29 June 2007'
__credits__   = '\n\n  Credits: ' + \
                '\n\n\t  tsLibGUI Import & Application Launch Features: ' + \
                '\n\t  Copyright (c) 2007-2009 Frederick A. Kier.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\n\t  Python Curses Module API & ' + \
                'Run Time Library Features:' + \
                '\n\t  Copyright (c) 2001-2013 ' +\
                'Python Software Foundation.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  PSF License Agreement for Python 2.7.3 & 3.3.0' + \
                '\n\n\t  wxWidgets (formerly wxWindows) & ' + \
                'wxPython API Features:' + \
                '\n\t  Copyright (c) 1992-2008 Julian Smart, ' + \
                'Robert Roebling,' + \
                '\n\t\t\tVadim Zeitlin and other members of the ' + \
                '\n\t\t\twxWidgets team.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  wxWindows Library License' + \
                '\n\n\t  nCurses API & Run Time Library Features:' + \
                '\n\t  Copyright (c) 1998-2011 ' + \
                'Free Software Foundation, Inc.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  GNU General Public License, ' + \
                'Version 3, 29 June 2007'

__line1__ = '%s, v%s (build %s)' % (__title__, __version__, __date__)
__line2__ = 'Author(s): %s' % __authors__
__line3__ = '%s' % __copyright__

if len( __credits__) == 0:
    __line4__ = '%s' % __license__
else:
    __line4__ = '%s%s' % (__license__, __credits__)

__header__ = '\n\n%s\n\n  %s\n  %s\n  %s\n' % (__line1__,
                                               __line2__,
                                               __line3__,
                                               __line4__)

mainTitleVersionDate = __line1__

import tsWxGlobals as wx
##from tsWxEvent import EVT_CLOSE              # [X]-Button Click (Frame)
##from tsWxEvent import EVT_HELP               # [?]-Button Click (Dialog)
##from tsWxEvent import EVT_ICONIZE            # [_]-Button Click (Frame)
##from tsWxEvent import EVT_MAXIMIZE           # [Z]-Button Click (Frame)
##from tsWxEvent import EVT_RESTOREDOWN        # [z]-Button Click (Frame)
from tsWxEvent import EVT_COMMAND_LEFT_CLICK # [Left Mouse]-Button Click
from tsWxControl import Control
from tsWxPoint import Point as wxPoint
from tsWxRect import Rect as wxRect
from tsWxSize import Size as wxSize

#---------------------------------------------------------------------------

##DEBUG = True
DEBUG = wx.Debug_GUI_Launch | \
        wx.Debug_GUI_Progress | \
        wx.Debug_GUI_Termination | \
        wx.Debug_GUI_Exceptions

##VERBOSE = True
VERBOSE = wx.Debug_GUI_Configuration

#---------------------------------------------------------------------------

class Button(Control):
    '''
    A button is a control that contains a text string, and is one of the
    most common elements of a GUI. It may be placed on a dialog box or
    panel, or indeed almost any other window.
    '''
    def __init__(self,
                 parent,
                 id=wx.ID_ANY,
                 label=wx.EmptyString,
                 pos=wx.DefaultPosition,
                 size=wx.DefaultSize,
                 style=0,
                 validator=wx.DefaultValidator,
                 name=wx.ButtonNameStr,
                 useClientArea=True):
        '''
        Create and show a button. The preferred way to create standard buttons
        is to use a standard ID and an empty label. In this case wxWigets will
        automatically use a stock label that corresponds to the ID given.
        These labels may vary across platforms as the platform itself will
        provide the label if possible. In addition, the button will be
        decorated with stock icons under GTK+ 2.
        '''
        theClass = 'Button'

        wx.RegisterFirstCallerClassName(self, theClass)

        Control.__init__(self,
                         parent,
                         id=id,
                         pos=pos,
                         size=size,
                         style=style,
                         validator=validator,
                         name=name)

        # Capture initial caller parametsrs before they are changed
        self.caller_parent = parent
        self.caller_id = id
        self.caller_label = label
        self.caller_pos = pos
        self.caller_size = size
        self.caller_style = style
        self.caller_validator = validator
        self.caller_name = name

        self.tsBeginClassRegistration(theClass, id)

        if DEBUG:
            self.logger.debug('               self: %s' % self)
            self.logger.debug('             parent: %s' % parent)
            self.logger.debug('                 id: %s' % self.ts_Id)
            self.logger.debug('         AssignedId: %s' % self.ts_AssignedId)
            self.logger.debug('              label: %s' % label)
            self.logger.debug('                pos: %s' % str(pos))
            self.logger.debug('               size: %s' % str(size))
            self.logger.debug('              style: 0x%X' % style)
            self.logger.debug('               name: %s' % name)

        self.ts_Default = False

        self.ts_Name = name
        self.ts_Parent = parent

        theTopLevelClass = self
        self.SetTopLevelAncestor(theTopLevelClass)

        if parent is None:
            self.ts_GrandParent = None
        else:
            self.ts_GrandParent = parent.Parent

        if False:
            # Leaves artifacts of different color than parent background.
            self.ts_BackgroundColour = wx.ThemeToUse[
                'BackgroundColour'].lower()
            self.ts_ForegroundColour = wx.ThemeToUse[
                'ForegroundColour'].lower()
        else:
            # Inter-button spaces have same color as parent background.
            self.ts_BackgroundColour = self.Parent.ts_BackgroundColour
            self.ts_ForegroundColour = self.Parent.ts_ForegroundColour

        self.ts_Style = style

        self.ts_Label = label
        self.ts_ButtonText = self.tsStripAcceleratorTextLabel(label)

        self.ts_UseClientArea = useClientArea
        self.ts_Validator = validator

        (myRect, myClientRect) = self.tsButtonLayout(
            parent, pos, size, style, name)
        self.ts_Rect = myRect
        self.ts_ClientRect = myClientRect

        # Automatically Bind all mouse events ASAP (now).
        # Will register event in the SystemEventTable.
        event = EVT_COMMAND_LEFT_CLICK
        handler = self.tsOnLeftClick
        source = self
        self.Bind(event,
                  handler,
                  source,
                  useSystemEventTable=True)

        self.tsEndClassRegistration(theClass)

    #-----------------------------------------------------------------------

    def Create(self,
               parent,
               id=-1,
               label=wx.EmptyString,
               pos=wx.DefaultPosition,
               size=wx.DefaultSize,
               style=0,
               validator=wx.DefaultValidator,
               name=wx.ButtonNameStr,
               pixels=True):
        '''
        Create the GUI Button for 2-phase creation.
        '''
        if label == wx.EmptyString:
            # Prevent pylint warning.
            pass

        (myRect, myClientRect) = self.tsButtonLayout(
            parent, pos, size, style, name)
        self.ts_Rect = myRect
        self.ts_ClientRect = myClientRect

        self.ts_Handle = self.tsCursesNewWindow(myRect, pixels=pixels)

        return (self.ts_Handle is not None)

    #-----------------------------------------------------------------------

    @staticmethod
    def GetClassDefaultAttributes(variant=wx.WINDOW_VARIANT_NORMAL):
        '''
        Get the default attributes for this class. This is useful if you want
        to use the same font or colour in your own control as in a standard
        control -- which is a much better idea than hard coding specific
        colours or fonts which might look completely out of place on the
        user system, especially if it uses themes.

        The variant parameter is only relevant under Mac currently and is
        ignored under other platforms. Under Mac, it will change the size of
        the returned font. See wx.Window.SetWindowVariant for more about this.
        '''
        pass

    #-----------------------------------------------------------------------

##    TBD - Should be staticmethod
##    @staticmethod
    def GetDefaultSize(self):
        '''
        Returns the default button size for this platform.
        '''
        thickness = self.tsIsBorderThickness(
            style=self.ts_Style, pixels=False)

        if False:
            width = (len(self.ts_ButtonText) + 1) * wx.pixelWidthPerCharacter

            height = (1) * wx.pixelHeightPerCharacter
        else:
            width = (thickness.width + \
                     len(self.ts_ButtonText) + 1) * wx.pixelWidthPerCharacter

            height = (thickness.height + 1) * wx.pixelHeightPerCharacter

        return (wxSize(width, height))

    #-----------------------------------------------------------------------

    def SetDefault(self):
        '''
        This sets the button to be the default item for the panel or
        dialog box.
        '''
        self.ts_Default = True

    #-----------------------------------------------------------------------
    # Begin tsWx API Extensions

    def tsButtonLayout(self, parent, pos, size, style, name):
        '''
        Calculate position and size of button based upon arguments.
        '''
        if (not (self.ts_Handle is None)):

            # Inhibit operation once self.ts_Handle has been
            # assigned a curses window identifier.
            return (self.ts_Rect, self.ts_ClientRect)

        if style is None:
            # Prevent pylint warning.
            pass

        thePosition = self.tsGetClassInstanceFromTuple(pos, wxPoint)
        theSize = self.tsGetClassInstanceFromTuple(size, wxSize)

        theDefaultPosition = self.tsGetClassInstanceFromTuple(
            wx.DefaultPosition, wxPoint)
 
        theDefaultSize = self.tsGetClassInstanceFromTuple(
            wx.DefaultSize, wxSize)

        label = self.ts_ButtonText

        # Ignore UseClientArea because cursor wraps beyond bottom
        # when last character has been output.
        if self.ts_UseClientArea:
            offset = wxPoint(0, 0)
        else:
            offset = wxPoint(0, 0)

        if theSize == theDefaultSize:
            # theDefaultSize

            theLabelSize = theSize
            self.logger.debug(
                '      theLabelSize (begin): %s' % theLabelSize)

            if label is wx.EmptyString:
                # TBD - Adjust for cursor width
                theLabelSize = wxSize(
                    (len(label) + len('[]')) * wx.pixelWidthPerCharacter,
                    wx.pixelHeightPerCharacter)

            elif label.find('\n') == -1:
                # TBD - Remove adjustment for extra cursor width
                theLabelSize = wxSize(
                    (len(label) + len('[ ]')) * wx.pixelWidthPerCharacter,
                    wx.pixelHeightPerCharacter)
            else:
                # TBD - Remove adjustment for extra cursor width
                theLines = label.split('\n')
                maxWidth = 0
                maxHeight = len(theLines)
                for aLine in theLines:
                    if len(aLine) > maxWidth:
                        maxWidth = len(aLine)
                theLabelSize = wxSize(
                    (maxWidth + len('[ ]')) * wx.pixelWidthPerCharacter,
                    maxHeight * wx.pixelHeightPerCharacter)

            self.logger.debug(
                '      theLabelSize (end): %s' % theLabelSize)

            if thePosition == theDefaultPosition:
                # theDefaultPosition

                myRect = wxRect(parent.ClientArea.x,
                                parent.ClientArea.y,
                                theLabelSize.width,
                                theLabelSize.height)

            else:
                # Not theDefaultPosition

                myRect = wxRect(offset.x + thePosition.x,
                                offset.y + thePosition.y,
                                theLabelSize.width,
                                theLabelSize.height)

        else:
            # Not theDefaultSize

            if thePosition == theDefaultPosition:
                # theDefaultPosition

                myRect = wxRect(parent.ClientArea.x,
                                parent.ClientArea.y,
                                theSize.width,
                                theSize.height)

            else:
                # Not theDefaultPosition

                myRect = wxRect(offset.x + thePosition.x,
                                offset.y + thePosition.y,
                                theSize.width,
                                theSize.height)

##        theDisplayRect = self.Display.GetClientArea()
##        if not theDisplayRect.InsideRect(myRect):
##            myRect = wxRect(
##                max(thePosition.x, theDisplayRect.x),
##                max(thePosition.y, theDisplayRect.y),
##                min(theSize.width, theDisplayRect.width),
##                min(theSize.height, theDisplayRect.height - 1))

##        if not theDisplayRect.InsideRect(myRect):
##            myRect = theDisplayRect

        myClientRect = wxRect(myRect.x,
                              myRect.y,
                              myRect.width,
                              myRect.height)

        self.tsTrapIfTooSmall(name, myRect)
        fmt = 'parent=%s; pos=%s; size=%s; ' + \
              'name="%s"; label="%s"; myRect=%s'
        msg = fmt % (parent, str(pos), str(size), name, label, str(myRect))

        self.logger.debug('    tsButtonLayout: %s' % msg)

        return (myRect, myClientRect)

    #-----------------------------------------------------------------------

    def tsOnLeftClick(self, evt):
        '''
        '''
        triggeringObject = self
        objectId = triggeringObject.ts_AssignedId

        objectCriteria = 'EVT_COMMAND_LEFT_CLICK for ' + \
                         '[%s]-Button' % self.GetLabel()
        triggeringEvent = EVT_COMMAND_LEFT_CLICK
 
##        if self.GetLabel().find('CloseButton') > -1:

##            objectCriteria = 'EVT_CLOSE for [X]-Button'
##            triggeringEvent = EVT_CLOSE

##        elif self.GetLabel().find('HelpButton') > -1:

##            objectCriteria = 'EVT_HELP for [?]-Button'
##            triggeringEvent = EVT_HELP

##        elif self.GetLabel().find('MaximizeButton') > -1:

##            objectCriteria = 'EVT_MAXIMIZE for [Z]-Button'
##            triggeringEvent = EVT_MAXIMIZE

##        elif self.GetLabel().find('MinimizeButton') > -1:

##            objectCriteria = 'EVT_ICONIZE for [_]-Button'
##            triggeringEvent = EVT_ICONIZE

##        elif self.GetLabel().find('RestoreButton') > -1:

##            objectCriteria = 'EVT_RESTOREDOWN for [z]-Button'
##            triggeringEvent = EVT_RESTOREDOWN

##        else:

##            objectCriteria = 'EVT_COMMAND_LEFT_CLICK for ' + \
##                             '[%s]-Button' % self.GetLabel()
##            triggeringEvent = EVT_COMMAND_LEFT_CLICK

####        triggeringObject.ProcessEvent(triggeringEvent)

        results = self.tsProcessEventTables(
            objectCriteria=objectCriteria,
            objectId=objectId,
            triggeringEvent=triggeringEvent,
            triggeringObject=triggeringObject)

    #-----------------------------------------------------------------------

    def tsShow(self):
        '''
        Create and update Button specific native GUI.
        '''
        if self.tsIsShowPermitted:

            if self.ts_Handle is None:

                self.Create(self.Parent,
                            id=self.ts_AssignedId,
                            label=self.ts_Label,
                            pos=self.ts_Rect.Position,
                            size=self.ts_Rect.Size,
                            style=self.ts_Style,
                            validator=self.ts_Validator,
                            name=self.Name,
                            pixels=True)

            try:
                self.tsRegisterShowOrder()
            except Exception as e:
                self.logger.error('%s.tsShow; %s' % (__title__, e))

            self.tsUpdate()

    #-----------------------------------------------------------------------

    def tsUpdate(self):
        '''
        Draw the actual features of the Button.
        '''
        if self.tsIsShowPermitted:

            self.tsCursesScrollOk(0)
            self.tsCursesBkgd()
            self.tsCreateBorder()
            self.tsCreateButton()

    # End tsWx API Extensions
    #-----------------------------------------------------------------------
#---------------------------------------------------------------------------
 
if __name__ == '__main__':

    print(__header__)
